### ISSUE TITLE
Implementation of the Proximal Duality Gap Metric

### ISSUE THREAD

**Issue by rybo4490** [proximal-dg/proximal_dg#1]
Hi,

I was trying to implement this metric to monitor my WGAN convergence, but the algorithm in the supplementary document and the code in proximal_duality_gap.py don't seem to match. Could you please clarify this? Thank you!

**Comment by sahilsid** [comment]
Hi @rybo4490,
Could you provide some more detail regarding which part you are having difficulty with? The code does follow the algorithm given in the supplementary material but is written in a generic way so that it can be applied to any GAN formulation. Computing proximal DG requires computing **V_{D_w}** and **V^{\lambda}_{G_w}**, this corresponds to **M1** and **M2** respectively in the code. As each of these involves training one player of a GAN using its loss function, you can see that there is a **Trainer** for each. The proximal objective is specified as the loss function for the generator trainer (an instance of ProximalGeneratorLoss).

**Comment by rybo4490** [comment]
@sahilsid 
Thank you for the quick response! I couldn't find the ProximalGeneratorLoss method in the code earlier, so I tried implementing my own method for the Proximal Duality Gap. I'll verify it with the code I saw just now and get back to you if I still have issues! Thanks again

**Comment by rybo4490** [comment]
@sahilsid So just three quick clarifications and this will make complete sense to me! First, are you using just a regular Discriminator Loss  (For example Wasserstein Discriminator Loss) for M1 (V_D_w) and then ProximalGeneratorLoss for M2? I'm just not able to see where you pass proximal_evaluation_loss in the code. Second, is the perturbation of the model parameters before the Duality Gap estimation necessary? I don't see it in the algorithm. Lastly, are you subtracting the Discriminator loss and the Generator Loss for calculating the Duality Gap? That's what it looks like in the code. Thanks again for the clarification and I really appreciate all the help!

**Comment by sahilsid** [comment]
@rybo4490 
1. Yes, V_D_w (M1) is obtained using the original objective V, and not V^{\lambda}. Ideally, optimizing D w.r.t V or V^{\lambda} would be the same as an optimal D* for V would be locally optimal in its neighborhood, implying it is also optimal for V^{\lambda}. The proximal objective is used for training the generator to obtain V^{\lambda}_{G_w} (M2).
Presently, evaluation_loss and proximal_evaluation_loss are the same (the regular discriminator loss) because, during M2 computation, we are storing the optimal discriminator for the proximal objective, to reuse and reduce computational redundancy. 
2.  The perturbation in the model parameters is not necessary but can be helpful in overcoming saddle points where M1 / M2 computation may get stuck and affect the DG estimate obtained since we are using gradient descent. This is explained more in [this work](https://arxiv.org/abs/2012.06723).
3. We are not subtracting the discriminator and generator losses. We are evaluating the objective function (equals negative of discriminator loss) at (a) the worst-case discriminator and (b) the worst-case generator obtained through their corresponding optimization. The difference between them is the duality gap.
Hope this helps. Thanks

**Comment by rybo4490** [comment]
Thanks a bunch! That clears up everything.
